lib/block_view.go

```diff
@@ -7060,6 +7342,153 @@ func (bav *UtxoView) _connectSwapIdentity(
 	return totalInput, totalOutput, utxoOpsForTxn, nil
 }
 
+// _verifyAccessSignature verifies if the accessSignature is correct. Valid
+// accessSignature is the signed hash of (derivedPublicKey + expirationBlock)
+// in DER format, made with the ownerPublicKey.
+func _verifyAccessSignature(ownerPublicKey []byte, derivedPublicKey []byte,
+	expirationBlock uint64, accessSignature []byte) error {
+
+	// Compute a hash of derivedPublicKey+expirationBlock
+	expirationBlockByte := UintToBuf(expirationBlock)
+	accessByte := append(derivedPublicKey, expirationBlockByte[:]...)
+	accessHash := Sha256DoubleHash(accessByte)
+
+	// Convert ownerPublicKey to *btcec.PublicKey
+	if len(ownerPublicKey) != btcec.PubKeyBytesLenCompressed {
+		fmt.Errorf("_verifyAccessSignature: Problem parsing owner public key")
+	}
+	ownerPk, err := btcec.ParsePubKey(ownerPublicKey, btcec.S256())
+	if err != nil {
+		return errors.Wrapf(err, "_verifyAccessSignature: Problem parsing owner public key: ")
+	}
+
+	// Convert accessSignature to *btcec.Signature
+	signature, err := btcec.ParseDERSignature(accessSignature, btcec.S256())
+	if err != nil {
+		return errors.Wrapf(err, "_verifyAccessSignature: Problem parsing access signature: ")
+	}
+
+	// Verify signature
+	if !signature.Verify(accessHash[:], ownerPk) {
+		return fmt.Errorf("_verifyAccessSignature: Invalid signature")
+	}
+
+	return nil
+}
+
+func (bav *UtxoView) _connectAuthorizeDerivedKey(
+	txn *MsgBitCloutTxn, txHash *BlockHash, blockHeight uint32, verifySignatures bool) (
+	_totalInput uint64, _totalOutput uint64, _utxoOps []*UtxoOperation, _err error) {
+
+	// Check that the transaction has the right TxnType.
+	if txn.TxnMeta.GetTxnType() != TxnTypeAuthorizeDerivedKey {
+		return 0, 0, nil, fmt.Errorf("_connectAuthorizeDerivedKey: called with bad TxnType %s",
+			txn.TxnMeta.GetTxnType().String())
+	}
+
+	txMeta := txn.TxnMeta.(*AuthorizeDerivedKeyMetadata)
+
+	// Validate the operation type.
+	if txMeta.OperationType != AuthorizeDerivedKeyOperationValid &&
+		txMeta.OperationType != AuthorizeDerivedKeyOperationNotValid {
+		return 0, 0, nil, fmt.Errorf("_connectAuthorizeDerivedKey: called with bad OperationType %s",
+			txn.TxnMeta.GetTxnType().String())
+	}
+
+	// Make sure transaction hasn't expired.
+	if txMeta.ExpirationBlock <= uint64(blockHeight) {
+		return 0, 0, nil, RuleErrorAuthorizeDerivedKeyExpiredDerivedPublicKey
+	}
+
+	// Validate the owner public key.
+	ownerPublicKey := txn.PublicKey
```

Comment:
Fun question for you. I don't think it's harmful, but can you have a derived key off of a derived key? E.g:

`original_owner_pubkey -> derived_key_1 -> derived_key_2` ?

Reply:
Hah nice. You can definitely have a derived key of a derived key. Though `key_2` won't automatically sign on behalf of `owner_pubkey`; you need an access signature and a new authorize transaction for that.

Reply:
A little sus but seems harmless so resolving.